File: src/main/java/com/keboola/jdbc/meta/SchemaCache.java
package com.keboola.jdbc.meta;

import com.keboola.jdbc.http.StorageApiClient;
import com.keboola.jdbc.http.model.Bucket;
import com.keboola.jdbc.http.model.TableInfo;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Collections;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Caches Storage API metadata (buckets and tables) with a configurable TTL.
 * Thread-safe implementation using ConcurrentHashMap.
 */
public class SchemaCache {

    private static final Logger LOG = LoggerFactory.getLogger(SchemaCache.class);
    private static final long DEFAULT_TTL_MS = 60_000; // 60 seconds

    private final StorageApiClient storageClient;
    private final long ttlMs;

    private volatile List<Bucket> cachedBuckets;
    private volatile long bucketsTimestamp;

    private final ConcurrentHashMap<String, CachedEntry<List<TableInfo>>> tablesCache = new ConcurrentHashMap<>();

    public SchemaCache(StorageApiClient storageClient) {
        this(storageClient, DEFAULT_TTL_MS);
    }

    public SchemaCache(StorageApiClient storageClient, long ttlMs) {
        this.storageClient = storageClient;
        this.ttlMs = ttlMs;
    }

    /**
     * Returns cached buckets, refreshing if expired.
     */
    public List<Bucket> getBuckets() {
        long now = System.currentTimeMillis();
        if (cachedBuckets == null || (now - bucketsTimestamp) > ttlMs) {
            LOG.debug("Refreshing buckets cache");
            try {
                cachedBuckets = storageClient.listBuckets();
                bucketsTimestamp = now;
            } catch (Exception e) {
                LOG.warn("Failed to refresh buckets cache: {}", e.getMessage());
                if (cachedBuckets == null) {
                    return Collections.emptyList();
                }
                // Return stale data on refresh failure
            }
        }
        return cachedBuckets;
    }

    /**
     * Returns cached tables for a bucket, refreshing if expired.
     */
    public List<TableInfo> getTables(String bucketId) {
        long now = System.currentTimeMillis();
        CachedEntry<List<TableInfo>> entry = tablesCache.get(bucketId);

        if (entry == null || (now - entry.timestamp) > ttlMs) {
            LOG.debug("Refreshing tables cache for bucket: {}", bucketId);
            try {
                List<TableInfo> tables = storageClient.listTables(bucketId);
                tablesCache.put(bucketId, new CachedEntry<>(tables, now));
                return tables;
            } catch (Exception e) {
                LOG.warn("Failed to refresh tables cache for bucket {}: {}", bucketId, e.getMessage());
                if (entry == null) {
                    return Collections.emptyList();
                }
                return entry.data;
            }
        }
        return entry.data;
    }

    /**
     * Invalidates all cached data.
     */
    public void invalidate() {
        cachedBuckets = null;
        bucketsTimestamp = 0;
        tablesCache.clear();
    }

    /**
     * Invalidates cached tables for a specific bucket.
     */
    public void invalidateBucket(String bucketId) {
        tablesCache.remove(bucketId);
    }

    private static class CachedEntry<T> {
        final T data;
        final long timestamp;

        CachedEntry(T data, long timestamp) {
            this.data = data;
            this.timestamp = timestamp;
        }
    }
}

File: vscode-sqltools/src/ls/schema-cache.ts
import { SCHEMA_CACHE_TTL_MS } from '../constants';

/**
 * Cached entry wrapping data and its fetch timestamp.
 */
interface CachedEntry<T> {
  data: T;
  timestamp: number;
}

/**
 * Caches Storage API metadata (buckets and tables) with a configurable TTL.
 *
 * Within the TTL window, cached data is returned without calling the fetch
 * function. After TTL expiry, the fetch function is called to refresh. If the
 * fetch fails and stale data is available, the stale data is returned as a
 * fallback. If no stale data exists (first call failure), the error propagates.
 *
 * No locking is needed because Node.js is single-threaded.
 */
export class SchemaCache {
  private readonly ttlMs: number;

  private bucketsEntry: CachedEntry<any[]> | null = null;
  private tablesEntries = new Map<string, CachedEntry<any[]>>();

  constructor(ttlMs: number = SCHEMA_CACHE_TTL_MS) {
    this.ttlMs = ttlMs;
  }

  /**
   * Returns cached buckets, refreshing via fetchFn if the cache is expired.
   * On fetch error: returns stale data if available, throws if not.
   */
  async getBuckets<T>(fetchFn: () => Promise<T[]>): Promise<T[]> {
    const now = Date.now();

    if (this.bucketsEntry && (now - this.bucketsEntry.timestamp) <= this.ttlMs) {
      return this.bucketsEntry.data as T[];
    }

    try {
      const data = await fetchFn();
      this.bucketsEntry = { data, timestamp: Date.now() };
      return data;
    } catch (err) {
      if (this.bucketsEntry) {
        // Return stale data on refresh failure
        return this.bucketsEntry.data as T[];
      }
      throw err;
    }
  }

  /**
   * Returns cached tables for a bucket, refreshing via fetchFn if the cache
   * is expired. On fetch error: returns stale data if available, throws if not.
   */
  async getTables<T>(bucketId: string, fetchFn: () => Promise<T[]>): Promise<T[]> {
    const now = Date.now();
    const entry = this.tablesEntries.get(bucketId);

    if (entry && (now - entry.timestamp) <= this.ttlMs) {
      return entry.data as T[];
    }

    try {
      const data = await fetchFn();
      this.tablesEntries.set(bucketId, { data, timestamp: Date.now() });
      return data;
    } catch (err) {
      if (entry) {
        // Return stale data on refresh failure
        return entry.data as T[];
      }
      throw err;
    }
  }

  /**
   * Invalidates all cached data (buckets and all per-bucket tables).
   */
  invalidate(): void {
    this.bucketsEntry = null;
    this.tablesEntries.clear();
  }

  /**
   * Invalidates cached tables for a specific bucket.
   */
  invalidateBucket(bucketId: string): void {
    this.tablesEntries.delete(bucketId);
  }
}
